### PR TITLE
Introduce SearchResponseSections base class

### PR DESCRIPTION
**Pull request by javanna**
`SearchResponseSections` is the common part extracted from `InternalSearchResponse` that can be shared between high level REST and elasticsearch. The only bits left out are around serialization which is not supported. This way it can accept `Aggregations` as a constructor argument, without requiring `InternalAggregations`, as the high level REST client uses its own objects for aggs parsing rather than internal ones.

This change also makes Aggregations implement ToXContent, and Aggregation extend ToXContent. Especially the latter is suboptimal but the best solution that allows to share as much code as possible between core and the client, that doesn't require messing with generics and making the api complicated. Also it doesn't have downsides as all of the current implementations of Aggregation do implement ToXContent already.

The idea is that with this change we can go back to #22533, add `fromXContent` to `SearchResponse`, where we will be able to return a search response created by passing in an instance of `SearchResponseSections` rather than `InternalSearchResponse`. `SearchResponseSections` holds `Aggregations` rather than `InternalAggregations`, which allows the high level REST client not to use `InternalAggregation` instances for aggs parsing, rather `ParsedAggregation`

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java [hunk 1]
```diff
@@ -98,4 +107,35 @@ public final boolean equals(Object obj) {
     public final int hashCode() {
         return Objects.hash(getClass(), aggregations);
     }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (aggregations.isEmpty()) {
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java]: `aggregations` is not initialized with emptyList() anymore, I think this might cause an issue here

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java [hunk 2]
```diff
@@ -19,148 +19,50 @@
...
 
 import java.io.IOException;
-import java.util.Collections;
-import java.util.Map;
 
-public class InternalSearchResponse implements Streamable, ToXContent {
+/**
+ * {@link SearchResponseSections} subclass that can be serialized over the wire.
+ */
+public class InternalSearchResponse extends SearchResponseSections implements Writeable, ToXContent {
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: `Writeable` is not necessary here

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java [hunk 3]
```diff
@@ -19,148 +19,50 @@
...
-    private boolean timedOut;
-
-    private Boolean terminatedEarly = null;
-
-    private int numReducePhases = 1;
-
-    private InternalSearchResponse() {
-    }
-
     public InternalSearchResponse(SearchHits hits, InternalAggregations aggregations, Suggest suggest,
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: Do you think we could add some `@Nullable` here?

**Review by tlrx — COMMENTED** [review]
This looks nice, I left some comments/questions

File: core/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Base class that holds the various sections which a search response is
+ * composed of (hits, aggs, suggestions etc.) and allows to retrieve them.
+ * It implements {@link Writeable} as it is required by {@link SearchResponse},
+ * but this implementation doesn't support serialization over the wire, for which it should be sub-classed.
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponseSections.java]: the subtle difference there is that if you try to serialize the parsed search response by calling writeTo you get back a class cast exception rather than an unsupported operation exception. I think it's ok to create internal search response in readFrom as we can control that. I think I will try to have the inner class not implement Writeable, yet leave the writeTo method that throws exception there.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 5]
```diff
@@ -39,14 +39,13 @@
...
 import static org.elasticsearch.action.search.ShardSearchFailure.readShardSearchFailure;
-import static org.elasticsearch.search.internal.InternalSearchResponse.readInternalSearchResponse;
 
 /**
  * A response of a search request.
  */
 public class SearchResponse extends ActionResponse implements StatusToXContentObject {
 
-    private InternalSearchResponse internalResponse;
+    private SearchResponseSections internalResponse;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: BaseSearchResponse makes me think that it has to do with its ancestor. I went for sections because it holds the different sections of the search response, in the same spirit as Aggregations.

File: core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java]: yea I forgot to push a commit. I had already fixed this, good catch.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java (hunk 3 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: Maybe, I tried to limit the unrelated changes here as it will go in our branch. Does the need for Nullable come from my changes?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java (hunk 2 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: I left it as SearchREsponseSections doesn't implement Writeable anymore

**Comment by javanna** [comment]
can you have another look @tlrx ?

File: core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java [hunk 6]
```diff
@@ -98,4 +107,36 @@ public final boolean equals(Object obj) {
...
+     * Directly write all the aggregations without their bounding object. Used by sub-aggregations (non top level aggs)
+     */
+    public XContentBuilder toXContentInternal(XContentBuilder builder, Params params) throws IOException {
+        for (Aggregation aggregation : aggregations) {
+            aggregation.toXContent(builder, params);
+        }
+        return builder;
+    }
+
+    //TODO add tests for this method
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java]: would you mind if I add tests as a follow-up so we can get this in rather quickly?

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java (hunk 5 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: Ok

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java (hunk 2 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: Sure, thanks. I prefer this, SearchResponseSections not declaring that it implements Writeable but still providing a writeTo method.

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java (hunk 3 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java]: > Does the need for Nullable come from my changes?

No, just thinking that it would be useful here.

**Review by tlrx — APPROVED** [review]
LGTM, thanks

File: core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java (hunk 6 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/Aggregations.java]: ++
